runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5278,6 +5287,40 @@ class INTERPRETER_CLASS
 			} else if (J9NtcPointer == argType) {
 				/* ffi_call expects the address of the pointer is the address of the stackslot. */
 				pointerValues[i] = (U_64)ffiArgs[i];
+#if JAVA_SPEC_VERSION >= 22
+				if (J9_ARE_ALL_BITS_SET(pointerValues[i], J9_FFI_DOWNCALL_HEAP_ARGUMENT_ID)) {
+					U_64 heapOffset = *(U_64 *)J9JAVAARRAY_EA(_currentThread,
+							J9_JNI_UNWRAP_REFERENCE(_sp + 10), /* The heap offset array */
+							curHeapArgIdx,
+							U_64);
+					j9object_t heapBase = (j9object_t)J9JAVAARRAYOFOBJECT_LOAD(_currentThread,
+							J9_JNI_UNWRAP_REFERENCE(_sp + 11), /* The heap base array */
+							curHeapArgIdx);
+					if (NULL == heapBase) {
+						rc = THROW_NPE;
+						goto done;
+					}
+					/* Extract the heap address from the base object of the heap segment which is
+					 * created by MemorySegment.ofArray(), MemorySegment.ofBuffer() or other methods
+					 * (which only returns part of the heap array) in OpenJDK.
+					 *
+					 * Note:
+					 * 1) Only MemorySegment objects representing the on-heap arrays reach here.
+					 * 2) Given the default offset of the original heap array is 8 bytes intended
+					 *    for the header, 8 bytes is subtracted from the offset of the passed-in
+					 *    heap segment so as to skip over the header (8 bytes) of the heap array
+					 *    to the content of the array against the offset.
+					 */
+					pointerValues[i] = (UDATA)(intptr_t)J9JAVAARRAY_EA(_currentThread, heapBase, heapOffset - 8, U_64);
+					curHeapArgIdx += 1;
+					if (!isHeapPassed) {
+						/* Set the flag to obtain the VMAccess so as to prevent the GC from
+						 * updating the heap address during the critical downcall.
+						 */
+						isHeapPassed = TRUE;
+					}
```

Comment:
For a local, there's no need to test the flag before setting it.

Reply:
Fixed.